### DISCUSSION TITLE
Announcing DocArray v2 a Pydantic extension for Machine learning

### DISCUSSION THREAD

**Discussion by samsja** [pydantic/pydantic#4801] [Projects using Pydantic]
# DocArray v2

Hey all, I am one of the core maintainers of [docarray](https://github.com/docarray/docarray), a project under the Linux foundation, and today I am very happy to announce that we are releasing the first alpha version of DocArray v2: a rewrite of DocArray from scratch.

DocArray's original version **was** already compatible Pydantic but this new rewrite is **built upon Pydanitc**. Pydantic is amazing for data handling but we wanted to build something focused on handling multi modal data (image, audio, 3d Mesh ...) for machine learning. We can see this DocArray v2 as a extension of Pydantic for machine learning.

Our goal is to equip the python web development world with a performant tool to deal with the new era of multi modal machine learning based application.

## How DocArray v2 is built on top of Pydantic

One of the core concepts of DocArray is the `Document` that is used to represent complex nested multi modal data. We released months ago a [dataclass feature](https://docarray.jina.ai/fundamentals/dataclass/) similar to what pydantic offer that has been widely appreciated by the community. Since then we decided that this way of working with `Document` should be the default one. We kicked off our rewriting and naturally Pydantic turned out to be the tools on which we needed to build our library. 

In DocArray v2 a [`Document`](https://github.com/docarray/docarray/blob/7565b6d7cab64a123128329de5e24832026e58be/docarray/document/document.py#L15) **is** a Pydantic `Model`, so it includes data validation and parsing, serialization to json (and to protobuf), FastAPI integration ...

You can interact with Document the same way you do with Pydantic Model:

```python
from typing import Optional

from docarray import Document
from docarray.typing import NdArray, ImageUrl, Embedding

class Image(Document):
    image: Optional[NdArray[3, 224, 224]] # the expected shape is 3, 224 ,224
    url: ImageUrl
    embedding: Embedding
```

Like Pydantic we provide predefined types, but with machine learning and multi modalities in mind.

Our types also provide some data processing ability. For instance you can do:

```python
doc = Image(url='https://github.com/docarray/artwork/blob/main/icon/color/docarray-icon-color-reverse.png')
doc.image = doc.url.load() # this will download the image and convert it into a numpy ndarray
```

Actually, DocArray is more than just an extension of Pydantic, offering more data structures, integrations with vector databases, protobuf serialization, and much more. Read on to learn more!

## More info

You can find our rewrite announcement blog post [here](https://github.com/docarray/notes/blob/main/blog/01-announcement.md).
If you are interested in this project we would love to hear from you! We are building in the open and rely on your feedback and input. You can join is in our [GitHub discussions](https://github.com/docarray/docarray/discussions/categories/docarray-v2) to talk about it.

The project itself, and its readme, can be found [here](https://github.com/docarray/docarray/blob/feat-rewrite-v2/README.md).

**Comment by nmj18txstate** [comment]
I'm runing this locally, I have pydantic '2.5.2' installed. It seems that this version is not compatible with the docarray package. Please guide how to know which different version of pydantic, docarray is expecting which has the ROOT_KEY attribute.

Chat with PDF Documents : https://huggingface.co/spaces/thewise/Chat-w-PDF
This app allows you to chat with documents. You can upload a PDF file and ask questions about it. In the backround uses the ConversationalRetrival chain from langchain and Streamlit for UI.
{
"FileName":"ULMFiT.pdf"
"FileType":"application/pdf"
"FileSize":979417
}
ImportError: Could not import docarray python package. Please install it with pip install "langchain[docarray]"`.
```
Traceback: File "C:\Nityam\Git\Chat-w-PDF\chat-w-pdf-env\Lib\site-packages\streamlit\runtime\scriptrunner\script_runner.py", line 534, in _run_script
exec(code, module.dict)
File "C:\Nityam\Git\Chat-w-PDF\app.py", line 112, in
main()

File "C:\Nityam\Git\Chat-w-PDF\app.py", line 103, in main
agent = load_agent(file_path, api_key)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "C:\Nityam\Git\Chat-w-PDF\chat-w-pdf-env\Lib\site-packages\streamlit\runtime\caching\cache_utils.py", line 212, in wrapper
return cached_func(*args, **kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "C:\Nityam\Git\Chat-w-PDF\chat-w-pdf-env\Lib\site-packages\streamlit\runtime\caching\cache_utils.py", line 241, in call
return self._get_or_create_cached_value(args, kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "C:\Nityam\Git\Chat-w-PDF\chat-w-pdf-env\Lib\site-packages\streamlit\runtime\caching\cache_utils.py", line 267, in _get_or_create_cached_value
return self._handle_cache_miss(cache, value_key, func_args, func_kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "C:\Nityam\Git\Chat-w-PDF\chat-w-pdf-env\Lib\site-packages\streamlit\runtime\caching\cache_utils.py", line 321, in _handle_cache_miss
computed_value = self._info.func(*func_args, **func_kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "C:\Nityam\Git\Chat-w-PDF\app.py", line 42, in load_agent
agent = ConversationalResponse(file_path, api_key)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "C:\Nityam\Git\Chat-w-PDF\src\main.py", line 99, in init
self.vector_store = ingest_data(self.chunks, embeddings)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "C:\Nityam\Git\Chat-w-PDF\src\main.py", line 50, in ingest_data
vector_store = DocArrayInMemorySearch.from_documents(chunks, embeddings)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "C:\Nityam\Git\Chat-w-PDF\chat-w-pdf-env\Lib\site-packages\langchain\schema\vectorstore.py", line 420, in from_documents

File "C:\Nityam\Git\Chat-w-PDF\chat-w-pdf-env\Lib\site-packages\langchain\vectorstores\docarray\in_memory.py", line 67, in from_texts

File "C:\Nityam\Git\Chat-w-PDF\chat-w-pdf-env\Lib\site-packages\langchain\vectorstores\docarray\in_memory.py", line 38, in from_params

File "C:\Nityam\Git\Chat-w-PDF\chat-w-pdf-env\Lib\site-packages\langchain\vectorstores\docarray\base.py", line 29, in _check_docarray_import`
```

**Reply by samsja** [reply]
Hey @nmj18txstate this seems to be a langchain issue, could you maybe open an issue there ? Which version of docarray and langchain are you using ?
